[#2] 회원가입, 로그인

2eve2 opened the pull request.

dding94 committed:
feat: 회원가입 기능 추가

2eve2 committed:
feat: 유저 저장 추가

2eve2 committed:
feat: 유ì가입 회원가입 유저 저장

dding94 committed:
feat: 회원 검증 및 예외 추가

dding94 committed:
feat: 회원 검증 및 예외 추가

2eve2 committed:
Merge remote-tracking branch 'origin/feature/02-create-user' into feature/02-create-user

2eve2 committed:
feat: non information token으로 세션 인증 방식 구현

dding94 committed:
test: findByEmail 기능 테스트

dding94 committed:
Rename: 폴더 추가 및 이름 수정

dding94 committed:
Comment: 주석 삭제

dding94 committed:
feat: 회원가입시 중복체크

dding94 committed:
Refactor: component -> bean 등록 변경

2eve2 commented:
커밋 컨벤션 통일할 수 있도록 논의해야할 것 같습니다.
- 커밋 제목 대소문자 
- 커밋 제목 분류 
등

f-lab-bright reviewed (commented):
.idea 파일은 로컬 ide의 설정 파일이라 ignore 시켜주셔야 합니다.

f-lab-bright commented on livecommerce/src/main/java/com/flab/livecommerce/application/UserLoginProcessor.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public User execute(UserLoginRequest userLoginRequest) {
+        User user = userRepository.findByEmail(userLoginRequest.getEmail());
+
+        if (null == user) {
+            throw new RuntimeException();
+        }
+
+        return user;
```
login 시에 패스워드 검증해야 하지 않나요?

2eve2 replied:
패스워드 검증 파트를 잊었네요! 
추가하도록 하겠습니다!

f-lab-bright commented on livecommerce/src/main/java/com/flab/livecommerce/application/UserCreateProcessor.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public void execute(UserCreateRequest userCreateRequest) {
+        User user = new User(
+            userCreateRequest.getEmail(),
+            userCreateRequest.getPassword(),
+            userCreateRequest.getNickname()
+        );
+
+        User findUser = userRepository.findByEmail(user.getEmail());
+        //중복 회원 검사
+        if (null != findUser) {
+            throw new RuntimeException();
+        }
+
+        userRepository.save(user);
```
패스워드를 평문으로 그냥 저장해도 괜찮을까요??
서비스에서는 어떻게 하고 있을까요??

f-lab-bright replied:
요거 찾아보셨나요?? :)

dding94 replied:
시큐리티를 사용하지않으면 SHA256과 SHA512 방법을 사용하는 것 같습니다. 그 중 대채로 원본 문자열에 임의의 값(salt)를 합치고 SHA 256 알고리즘을 적용시키는 것 같습니다.

2eve2 replied:
우선 비밀번호를 평문으로 데이터베이스에 저장하는 방식은 DB에 접근 가능한 사람들에게 공개되어 유출될 우려가 있으므로 옳지 않습니다. 

SHA 256, SHA 512 와 같은 복호화가 불가능한 **해시함수**를 이용하여 암호화한 비밀번호 입력값을 DB에 저장하는 방식을 쓸 수 있습니다. 하지만, 입력값이 같으면 출력값이 동일한 해시함수 특성상 **rainbow table**으로 해시값을 돌려보고 비밀번호를 해킹할 수 있습니다.

이를 막기 위해 **salt(랜덤 문자열)**을 비밀번호를 해싱하기 전에 붙여서 해쉬 값을 무작위로 만들어 비밀번호를 암호화하는 방식을 사용하고 있습니다.

f-lab-bright replied:
`security:spring-security-crypto` 의존성 정도만 사용해서 암호화를 구현해보셔도 될 것 같네요 :)
혹은 직접 만들어보셔도 괜찮구요 :)

2eve2 committed:
chore: 폴더 정리, gitignore 수정

2eve2 committed:
chore: test 코드 checkstyle 제외 설정

f-lab-bright commented on src/main/java/com/flab/livecommerce/presentation/UserController.java:
```diff
@@ -0,0 +1,49 @@
+        return "ok";
+    }
+
+    @PostMapping("/login")
+    public ResponseEntity<User> login(
+        @RequestBody @Valid UserLoginRequest userLoginRequest,
+        HttpServletResponse response
+    ) {
+        User loginUser = userManager.login(userLoginRequest);
+        tokenAuthorization.createToken(userLoginRequest, response);
+
+        return new ResponseEntity<>(loginUser, HttpStatus.OK);
+    }
```
로그인 이후에 인가 처리가 잘 되는지 확인을 위한 테스트 api를 하나 만들어 보시면 좋을 것 같네요 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/TokenAuthorization.java:
```diff
@@ -0,0 +1,50 @@
+
+    private Map<String, Object> tokenStorage = new ConcurrentHashMap<>();
+
+    public void createToken(Object userInfo, HttpServletResponse response) {
+        String userToken = UUID.randomUUID().toString();
+        tokenStorage.put(userToken, userInfo);
+        Cookie cookie = new Cookie(TOKEN_COOKIE_NAME, userToken);
+
+        log.info("token={}", userToken);
+        log.info("cookieName={}", TOKEN_COOKIE_NAME);
+        response.addCookie(cookie);
+    }
+
+    public Object getUserByToken(HttpServletRequest request) {
+        Cookie tokenCookie = findCookie(request, TOKEN_COOKIE_NAME);
+
+        if (null == tokenCookie) {
+            return new RuntimeException();
+        }
+
+        return tokenStorage.get(tokenCookie.getValue());
+    }
```
쿠키로 토큰을 전달하고, 받기로 결정한 이유가 있나요??

2eve2 replied:
세션 로그인을 공부해봤을 때 쿠키로 세션id 등을 전달받는 자료가 많아서 선택했었습니다.
하지만 보안상 브라우저에서 쿠키를 사용하지 않도록 할 수 있고, 앱까지 고려해보았을 때는 쿠키를 관리할 필요가 없기 때문에 쿠키 방식이 아닌 헤더 방식으로 변경하겠습니다.

2eve2 committed:
Merge branch 'main' into feature/02-create-user

2eve2 committed:
style: code style 수정

메소드 네이밍 수정, missing java doc 설정 제거

2eve2 committed:
Merge remote-tracking branch 'origin/feature/02-create-user' into feature/02-create-user

2eve2 committed:
Merge remote-tracking branch 'origin/main' into feature/02-create-user

dding94 committed:
test: User 관련 테스트 코드 추가

회원가입, 패스워드 암호화, 회원생성 및 로그인 프로세서 테스트

dding94 committed:
chore: 사용하지 않는 코드 삭제

dding94 committed:
refactor: 회원가입, 로그인 기능 리팩토링

스프링 시큐리티 의존성 추가하여 패스워드 단방향 암호화 및 중복회원, 아이디 패스워드 검사 로직 추가

dding94 committed:
style: checkStyle 맞게 메서드,변수명 변경

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserCreateProcessor.java:
```diff
@@ -0,0 +1,33 @@
+
+import com.flab.livecommerce.domain.user.User;
+import com.flab.livecommerce.domain.user.UserRepository;
+import com.flab.livecommerce.presentation.request.UserCreateRequest;
```
이렇게 만들면 application layer가 presentation layer를 의존하게 되는 구조가 되어서 layered 구조에서 안티패턴입니다.
항상 레이어간의 의존성의 방향에 신경 써주세요

dding94 replied:
앞으로 이런부분 더 신경쓰겠습니다.
이 말씀이 결국 presentation 에서 request 를 받은걸 application 에서 command dto 를 만들어 presentation 미리 변환을 하고 넘기는 방법으로 하려고 하는데 괜찮은 방법이 될까요?

그리고 만약 User 와 command dto 가 앞으로도 계속 동일한 변수를 갖는 상황이더라도 따로 User 객체에서의 메시지 처리 로직을 생각하면 command 로 분리하는게 좋을 것 같긴한데 이것도 괜찮을까요?

f-lab-bright replied:
presentation이 application을 의존하는건 괜찮지만, 역방향은 나쁠 수 있습니다 :)
이 부분만 지켜진다면 어떤 방법이든 가능합니다 :)
apllication의 command dto를 presentation에서 그대로 사용 하는 것도 가능하고, 레이어별 dto를 만들고 변환해서 전달하는 것도 가능하구요 :)
적절하게 선택해서 구성해보시면 될 것 같습니다 :)

2eve2 replied:
layer 의존성을 고려하면 request를 처리하는 파트를 presentation 영역에서 하고, application 영역으로 넘겨주어야겠군요. 수정하겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserCreateProcessor.java:
```diff
@@ -0,0 +1,33 @@
+        //중복 회원 검사
+        if (null != findUser) {
+            throw new IllegalStateException();
+        }
```
이 중복검사를 먼저 해버리는게, 좀 더 효과적일 수 있겠네요.
패스워드 인코딩이나 객체를 하나 더 생성하는 코스트를 줄여줄 가능성이 더 있으니까요 :)

2eve2 replied:
이메일, 닉네임 중복검사를 컨트롤러측에서 getmapping 방식으로 따로 api를 만드는 것이 좋을까요? 
서비스도 고려해보니 원래 회원가입할 때 이메일, 닉네임 중복검사를 별도로 확인 후 통과되면 회원가입을 하는 방법을 주로 사용하기 때문에 따로 api로 빼야하나 생각중입니다.

2eve2 replied:
답변 옮김)
이메일 같은건 중복검사가 회원가입이 실제 전송되기 이전에 검사되어야 합니다 :)
ux가 좋은 서비스들을 보시면 이메일을 작성할 때 중복검사가 실시간으로 이루어지는걸 보실 수 있을거에요.
이런 플로우가 가능하려면 해당 이메일이 사용가능한지 검사해주는 api가 있어야 가능하겠네요 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserLoginProcessor.java:
```diff
@@ -0,0 +1,31 @@
+
+import com.flab.livecommerce.domain.user.User;
+import com.flab.livecommerce.domain.user.UserRepository;
+import com.flab.livecommerce.presentation.request.UserLoginRequest;
```
이 부분도 마찬가지네요.
의존성의 방향이 잘못 됐네요

dding94 replied:
시큐리티의 crypto의존해도 된다는 말씀이 gradle에서만 이군요. 잘못 이해했습니다. 반영해서 수정하도록 하겠습니다.

f-lab-bright replied:
구현 상세로서 사용하는것과 비즈니스 로직에서 직접 의존하는건 차이가 있습니다 :)
물론 실제로는 모든 구현상세를 추상화해서 사용할 필요가 있는건 아니지만, 지금 프로젝트 하는 과정에서는 최대한 많이 추상화 해보는 연습을 해보면 좋을 것 같습니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserCreateProcessor.java:
```diff
@@ -0,0 +1,33 @@
+import com.flab.livecommerce.domain.user.User;
+import com.flab.livecommerce.domain.user.UserRepository;
+import com.flab.livecommerce.presentation.request.UserCreateRequest;
+import org.springframework.security.crypto.password.PasswordEncoder;
```
구현 상세는 인프라의 영역이라 이 부분도 의존성이 잘 나눠지지 못한 부분이네요.
스프링 시큐리티 의존성이 아닌 구현으로 바꿔야 하면 이부분이 다 고쳐져야 할 수 있겠죠

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserLoginProcessor.java:
```diff
@@ -0,0 +1,31 @@
+import com.flab.livecommerce.domain.user.User;
+import com.flab.livecommerce.domain.user.UserRepository;
+import com.flab.livecommerce.presentation.request.UserLoginRequest;
+import org.springframework.security.crypto.password.PasswordEncoder;
```
마찬가지구요 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/TokenAuthorization.java:
```diff
@@ -0,0 +1,50 @@
+    private Map<String, Object> tokenStorage = new ConcurrentHashMap<>();
+
+    public void createToken(Object userInfo, HttpServletResponse response) {
+        String userToken = UUID.randomUUID().toString();
```
이 아이디를 생성하는 것도 여기서 구체적인 방법을 결정해버리면, 추후에 변경이 일어나야 할 때 이 코드가 직접 고쳐져야겠네요.
ocp를 위반하는 코드가 되겠네요 그럼 :)

dding94 replied:
Generator로 토큰생성되도록 만든 코드가 있긴한데 커밋을 안해서 바로 반영하겠습니다.

f-lab-bright replied:
요기 TokenGenerator 사용하는 코드로 변경이 안되어 있네요 :)

dding94 replied:
네 이부분은 다른 브런치에서 작업하려고 계획중입니다.

dding94 committed:
test: 테스트 코드 리팩토링

CreateProcessor,LoginProcessor 테스트 코드 수정 작업

dding94 committed:
feat: 토큰 생성 기능 추가

구체클래스로 UUID로 랜덤하게 생성해주는 클래스도 추가함

dding94 committed:
refactor: 회원가입, 로그인 리팩토링

application layer 에서 presentation layer 의존하는 코드 삭제 및 리팩토링

2eve2 committed:
refactor: 리뷰 반영

비밀번호 암호화 추상화

2eve2 committed:
feat: api response

response 반환 통일

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/config/ProcessorConfig.java:
```diff
@@ -4,6 +4,7 @@
+import com.flab.livecommerce.infrastructure.encryption.BCryptPasswordEncryption;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
```
혹시 PasswordEncoder인터페이스가 아닌 구체 클래스 BcryptPasswordEncoder를 사용한 이유가 궁금합니다.
해당 클래스를 PasswordEncoder 로 의존하도록 코드를 변경하면 더 변경에 자유로워 지지 않을까요?

2eve2 replied:
시큐리티를 쓸 때 해쉬 알고리즘도 분리하여 사용할 생각을 못했네요. 추후 알고리즘이 변경될 수 있으니 passwordEncoder를 의존하도록 변경해놓았습니다!

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/config/ProcessorConfig.java:
```diff
@@ -12,26 +13,29 @@
 @Configuration
 public class ProcessorConfig {
 
+    @Bean
+    private static PasswordEncoder passwordEncoder() {
+        return new BCryptPasswordEncoder();
+    }
+
+
```
해당 코드를 무엇때문에 Bean 으로 등록하신건지 이유를 알고싶습니다.
이 코드는 제가 외부에서 사용하지 않는 코드이긴 합니다.

2eve2 replied:
아 이거는 명구님이 추상화하시긴 전에 Bean을 빼먹으신걸 리팩토링하고 추상화후에는 사용하지 않는거라 빼야하는데 잊어버렸네요ㅋㅋ
없애겠습니다!

dding94 replied:
이거 Bean 만 없애야 하는데 다 지워버리셨더군요 ㅠㅠ

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/config/ProcessorConfig.java:
```diff
@@ -12,26 +13,29 @@
+            UserRepository userRepository
     ) {
-        return new UserCreateProcessor(userRepository, passwordEncoder());
+        return new UserCreateProcessor(userRepository, new BCryptPasswordEncryption(new BCryptPasswordEncoder()) {
+        });
```
위의 메서드에 BCryptPasswordEncoder 구체 타입으로 반환되는 메서드가 있는데 new BCryptPasswordEncoder 를 사용하실 필요가 있을까 생각이 듭니다.

제 생각은 만약 더 많은 빈에서 BCryptPasswordEncoder가 동일하게 사용이 된다고 하면 전부다 new 로 만드는건 낭비가 아닐까 생각이됩니다.

2eve2 replied:
Security PasswordEncoder가 어떤 구현체를 사용할건지도 config 내에서 정해주어야겠다고 생각해서 new로 만들었었는데 좋지 않은 것 같네요. 시큐리티를 쓰느냐, 안쓰고 다른 방식을 쓰느냐의 관점에서만 생각해서 해당 new BCryptPasswordEncoder는 삭제하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/encryption/BCryptPasswordEncryption.java:
```diff
@@ -0,0 +1,24 @@
+import com.flab.livecommerce.domain.user.encryption.PasswordEncryption;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+public class BCryptPasswordEncryption implements PasswordEncryption {
```
이부분은 추가로 해당 클래스명을 SecurityPasswordEncoder 같은 상위의 이름을 쓰면 괜찮을 것 같은데 어떨까요?

2eve2 replied:
네 네이밍을 더 상위개념으로 쓰면 더 좋겠네요! 반영하겠습니다.

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/encryption/BCryptPasswordEncryption.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Override
+    public String encrypt(String password) {
+        String encryptedPassword = bCryptPasswordEncoder.encode(password);
+        return encryptedPassword;
```
해당 코드는 한줄로 리팩토링 하는게 더 깔끔할 것 같습니다!

2eve2 committed:
refactor: 리뷰 반영

클래스 네이밍, 비밀번호 암호화 추상화, 불필요한 코드 제거

2eve2 committed:
refactor: 리뷰 반영

이메일 중복 체크 api 별도 생성

DJ-archive commented:
전체적으로 리뷰 반영해보았습니다. 리뷰 부탁드립니다.

[참고]
1. 토큰생성 파트는 현재 컨트롤러에서 사용되지 않는데 사용하여 warning이 떠서 임시 삭제했습니다. 인증 파트에서 구현하면 될 것 같습니다.
2. 닉네임 중복체크를 할 수 있다고 판단하여, userchecprocessor를 사용할 때 execute()가 아닌 checkEmail() 메소드로 네이밍했습니다.

2eve2 committed:
refactor: ApiResponse 수정

response getter 추가

dding94 committed:
refactor: 응답 api 리팩토링

dding94 committed:
refactor: 암호화 부분 리팩토링

패스워드 암호화가 아닌 더 상위의 개념인 암호화라는 이름 사용 및 코드 리팩토링

dding94 committed:
fix: 시큐리티 의존성 변경

security-start -> security-crypto 의존하도록 코드 변경,
스타터를 사용해서 필터가 걸려 401에러 해결

dding94 committed:
refactor: 코드 리팩토링 진행

사용하지 않는 코드 제거, 코드 형식 통일, 변수명 변경

dding94 committed:
test: 테스트 코드 리팩토링

주석 제거, 변경된 코드에도 테스트가 원할하게 진행되도록 코드 리팩토링

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserCreateProcessor.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public void execute(CreateCommand command) {
+        userRepository.save(
+            new User(
+                command.getEmail(),
+                passwordEncryption.encrypt(command.getPassword()),
+                command.getNickname()
+            )
+        );
+    }
```
이미 가입된 email인 경우나 nickname 등이 중복되지 않아야 할 때 예외를 제대로 전달해주려면 여기서 미리 검사하는 과정이 있는게 좋을 것 같네요 :)
이렇게만 해버리면 db에서 제약조건을 통해서만 예외가 발생할텐데, 이건 디비나 디비 연결과정에서 어떤걸 사용하냐에 따라서도 예외가 달라지기 때문에 문제가 생길 수 있겠네요 :)

dding94 replied:
이부분에 원래 검사로직이 있었는데 가입된 회원을 체크하는 API를 만들게 되면서 분리가 되어 빼게 됐습니다. 다시 생각해보니 중복체크를 안하고 회원가입을 누르는 경우가 있겠다고 생각이 들어 회원가입 API 에서도 검사하는 로직이 들어가는게 맞다고 생각되는데 그런 의미에서 얘기를 하신걸까요?

f-lab-bright replied:
중복검사 api가 있더라도 가입 시점에 체크는 하셔야 합니다 :)
왜 그런지 고민해보시면 좋겠네요 :)

dding94 replied:
실제로 중복검사 API 가 있더라도 회원가입을 누르는 경우가 있기 때문에 그렇다고 생각이 듭니다.

f-lab-bright replied:
그건 중요한건 아니고, 중복검사를 했더라도, 실제 회원가입이 이루어지는 시점과 중복검사를 클라이언트에서 하는 시점이 다르기 때문에 중복검사 했다고 정말 중복되지 않음을 보장할 수 없기 때문입니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/UserLoginProcessor.java:
```diff
@@ -0,0 +1,30 @@
+    public void execute(LoginCommand command) {
+        User user = userRepository.findByEmail(command.getEmail());
+
+        if (idPasswordCheck(command, user)) {
+            throw new IllegalStateException();
+        }
```
idPasswordCheck() 가 true 일 때 예외를 던지는게 좀 어색한 플로우이긴 하네요 :)
좀 더 읽기 편하려면 검증이 실패하는 경우에 예외가 던져지도록 하는게 더 바람직하지 않을까 합니다 :)

dding94 replied:
이 부분을 false로 할지 true 로 할지 엄청 고민했는데 ! 을 붙여 실패하는 경우에 예외를 던지도록 수정하도록 하겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/command/user/CreateCommand.java:
```diff
@@ -0,0 +1,15 @@
+
+@Getter
+@AllArgsConstructor
+public class CreateCommand {
+
+    private String email;
+    private String password;
+    private String nickname;
+
+}
```
무엇을 create 하는지가 명확하지 않을 수 있겠네요 :)
Processor를 만드는 유즈케이스별로 클래스를 만드는 구조를 하신다면, Processor 내에 이너클래스로 커맨드 클래스를 선언하는 것도 괜찮은 방법입니다 :)

2eve2 replied:
네이밍을 할 때 좀 더 구체적으로 할 수 있도록 고민하겠습니다.
커멘드를 다 분리하는 것보다 이너클래스로 선언하여 함께 관리하는 방법이 더 좋을 것 같네요! 수정하겠습니다

dding94 replied:
유즈케이스별로 클래스를 만들게 됐을 때 Processor 내에 이너클래스로 커맨드 클래스를 구현하는 방법이 좋은 이유가 클래스파일이 불필요 하게 많아지는 것을 줄일 수 있어서 괜찮은 방법이라고 얘기하신게 맞을까요?

f-lab-bright replied:
클래스 파일이 줄어드는 것 보다, 사용처가 명확히 한 곳이다보니 같은 위치에 있는게 가독성이 높아지는 효과가 있기도 합니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/application/facade/UserManager.java:
```diff
@@ -0,0 +1,40 @@
+        userLoginProcessor.execute(command);
+    }
+
+    public void checkEmailDuplicated(String email) {
+        userCheckProcessor.execute(email);
+    }
```
요런 간단한 구현은 facade에서 직접 구현하시는 것도 가능합니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/domain/user/Encryption.java:
```diff
@@ -0,0 +1,7 @@
+package com.flab.livecommerce.domain.user;
+
+public interface Encryption {
```
Encryption 이라는 인터페이스 이름은 적절하지 않을 것 같네요.
PasswordEncoder, PasswordEncrypter 등의 이름이 좀 더 명확할 수 있을 것 같네요 :)

dding94 replied:
PasswordEncoder는 이미 시큐리티에서도 있는 이름이라 PasswordEncrypter 로 쓰는게 좀 더 좋다고 생각이 듭니다. 
PasswordEncrypter 로 변경하도록 하겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/domain/user/User.java:
```diff
@@ -0,0 +1,13 @@
+
+@Getter
+@AllArgsConstructor
+public class User {
+
+    private String email;
+    private String password;
+    private String nickname;
+}
```
요 녀석이 entity가 된다면...
식별자가 하나 있어야 할 것 같네요 :)
`Entity` 가 무엇인지, 어떤 특징이 있는지 한 번 찾아보시면 좋을 것 같네요 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/generator/RandomTokenGenerator.java:
```diff
@@ -0,0 +1,12 @@
+import com.flab.livecommerce.domain.user.TokenGenerator;
+import java.util.UUID;
+
+public class RandomTokenGenerator implements TokenGenerator {
```
UUIDTokenGenerator가 좀 더 적절할 수 있겠네요

dding94 replied:
UUID 사용하니까 대문자 4개 연속사용했다고 checkStyle 에 걸려버리네요..

2eve2 replied:
UUID 대신 NonInfoTokenGenerator로 바꿔 토큰 생성 방식을 명확히 나타내도록 바꾸었습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/persistence/inmemory/InMemoryUserRepository.java:
```diff
@@ -0,0 +1,23 @@
+    private Map<String, User> map = new ConcurrentHashMap<>();
+
+    public User save(User user) {
+        map.put(user.getEmail(), user);
```
email 을 key로 저장하는게 적절할까요?? :)

2eve2 replied:
database를 고려해서 id와 같은 식별자로 key를 저장해야할까요?

f-lab-bright replied:
디비를 신경쓰지 않더라도, entity는 식별자를 가지는게 좋습니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/presentation/handler/UserValidExceptionHandler.java:
```diff
@@ -0,0 +1,30 @@
+
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public String processValidationError(MethodArgumentNotValidException e) {
+        BindingResult bindingResult = e.getBindingResult();
+
+        StringBuilder builder = new StringBuilder();
+        for (FieldError fieldError : bindingResult.getFieldErrors()) {
+            builder.append("[");
+            builder.append(fieldError.getField());
+            builder.append("](은)는 ");
+            builder.append(fieldError.getDefaultMessage());
+            builder.append("]" + "\n");
+        }
+
+        return builder.toString();
+    }
```
에러 응답으로도 적절한 json 응답이 내려가는게 좀 더 좋겠네요.

dding94 replied:
이부분은 일단 보류해 뒀는데 커스텀 에러 만들면서 수정하도록 하겠습니다.

f-lab-bright commented on src/main/java/com/flab/livecommerce/presentation/shared/ApiResponse.java:
```diff
@@ -0,0 +1,37 @@
+    @Getter
+    static class Error {
+
+        //http 상태코드
+        private String code;
```
Http status code를 여기서 알고 있을 필요가 있을까요??
Exception 별로 응답으로 전달되어야 할 status code는 어차피 고정될테니, 예외 처리하는 부분에서 결정해줄 수 있을 것 같네요 :)
그리고 http가 아닌 응답으로도 사용할 수 있다는 점을 감안하면 http status code 라는 context는 너무 구체적인 정보가 들어간 것 같네요 :)

dding94 replied:
저도 동일한 Google, Twilio, Foursquare 등에서 에러로 상태코드를 적어줘서 응답을 반환하도록 설계한 곳이 있어 이 형식을 따라 사용을 했습니다. 왜 이렇게 사용했는지는 좀 찾아보도록 하겠습니다.

f-lab-bright replied:
거기에서 사용하는 code는 http status code 보다 내부적으로 관리하는 에러 코드일 가능성이 더 높습니다 :)

dding94 replied:
![스크린샷 2022-08-11 오후 1 30 35](https://user-images.githubusercontent.com/21376853/184064995-b785ca60-bcde-44df-ba7a-dc339ceac185.png)

이 부분을 참고한건데 내부적으로 관리하는 에러코드랑 상태코드가 둘 다 포함이 되어있어서.. 혼동이 오네요

f-lab-bright replied:
그건 저 api 들의 선택일 뿐입니다.
구글이라고 다 좋은 선택만 하는 것도 아니고, 저라고 항상 정답의 선택을 하는 것도 아닙니다.
하지만 제가 만약 구현한다면, 응답 바디에 http status code는 들어갈 필요가 없다고 생각하기 때문에 넣지 않을테구요.
이걸 꼭 넣어야 한다면, common 모듈에 있는 ApiResponse가 직접 이 정보를 가지고 전달해주기 보다는, presentation layer의 객체로 한 번 더 래핑한 형태로 사용할 것 같네요.
common 모듈에 있는 클래스는 http status code 라는 context를 알 필요가 전혀 없으니까요 :)
이걸 rest 로만 전달할 presentation layer에 두는 클래스로 만든다면 http status code 정보를 가지고 있어도 어색하지 않겠지만요 :)

f-lab-bright commented on src/test/java/com/flab/livecommerce/application/UserTest.java:
```diff
@@ -0,0 +1,109 @@
+        assertThat(encoder.matches("Test1234", result)).isTrue();
+        assertThat(encoder.matches("Test1237", result)).isFalse();
+    }
+}
```
test 코드는 lint가 아예 동작하지 않나요 혹시..??
마지막 라인이...?

2eve2 replied:
테스트코드는 한글화 네이밍 등 몇몇 규칙들을 적용하기 힘들고 경우에 따라 테스트코드를 suppression으로 처리한 사례가 있어서 빼두었습니다. 하지만 테스트코드도 체크스타일로 관리를 하고, 경우에 따라 체크스타일 룰을 조금 수정하는게 더 좋을 것 같네요. 다음 작업부터는 테스트코드도 체크스타일 적용하겠습니다 :)

f-lab-bright commented on src/main/java/com/flab/livecommerce/infrastructure/TokenAuthorization.java:
```diff
@@ -0,0 +1,50 @@
+
+    private Map<String, Object> tokenStorage = new ConcurrentHashMap<>();
+
+    public void createToken(Object userInfo, HttpServletResponse response) {
```
userInfo도 Object로 전달받는건 위험할 수 있겠네요.
더 명확한 구체 클래스로 데이터를 전달받는게 더 좋겠어요 :)

dding94 committed:
refactor: 회원 프로세스 클래스 리팩토링

커맨드 클래스를 프로세스클래스의 이너클래스로 변경, 패스워드 체크 실패시 예외를 내도록 변경, 회원가입시 회원 중복체크 로직 추가, 매니저는 import 변경 됨

dding94 committed:
chore: 시큐리티 패스워드 이름 변경

dding94 committed:
chore: import 변경

유저 프로세서의 이너클래스 반환하도록 변경함

dding94 committed:
test: 리팩토링 된 코드 반영

dding94 committed:
chore: 클래스 생성, TODO 추가

식별자도 추가하고 수정하려다가 범위가 좀 커져버려 롤백했습니다..

dding94 committed:
refactor: ProcessorConfig 리팩토링

algorithm 메서드 이 부분도 빈으로 관리하여 하나의 객체만 반환하도록 하는게 더 효율적일 것 같아 수정 했습니다.

2eve2 committed:
refactor: 리뷰 반영

user entity 식별자 추가 및 repository userMap 저장방식 수정, repository findByEmail 메소드 수정

2eve2 committed:
fix: 이메일 중복 체크 에러

2eve2 committed:
refactor: naming

변수명은 명사, 보다 더 구체적인 클래스명으로 변경

2eve2 committed:
refactor: 리뷰 반영

UserCheckProcessor 클래스 제거 후 facade에 구현

2eve2 committed:
refactor: user 식별자 수정

동시성 이슈 고려하여 AtomicLong으로 수정

dding94 committed:
chore: 사용하지 않는 클래스 파일 삭제

dding94 committed:
feat: 토큰 저장 기능 추가

Adapter 패턴을 사용하여 다른 저장소로 변경해도 유연하게 대처 가능하도록 구현

dding94 committed:
feat: login 요청시 body 토큰 반환

토큰관련 의존성 추가, 로그인 요청시 토큰 생성과 저장, JSON 으로 토큰 값 반환 하도록 구현

dding94 committed:
test: 토큰관련 의존성 추가

DJ-archive commented on src/main/java/com/flab/livecommerce/presentation/UserController.java:
```diff
@@ -0,0 +1,44 @@
+
+    @PostMapping
+    public ApiResponse signUp(@RequestBody @Valid UserCreateRequest request) {
+        userManager.createUser(request.toCommand());
```
userManager.create안에 이메일 중복 체크 로직을 추가하는 방식으로 회원가입을 구현했었는데, 그 방식보다는 이미 위에서 구현한 userManager.checkEmailDuplicated(email)을 controller에서 나타내주면 어떨까요? 코드가 중복되기도 하고 비즈니스 로직을 분리해서 보여주는 것이 더 좋아보이네요

DJ-archive commented on src/main/java/com/flab/livecommerce/application/UserLoginProcessor.java:
```diff
@@ -0,0 +1,56 @@
+            throw new IllegalStateException();
+        }
+
+        //TODO 토큰 만료 LocalDateTime 을 어디에 추가할지, 로그인 요청이 계속 들어오면?
```
현재는 login 서비스(프로세서) 로직 안에 토큰 생성과 저장 로직을 넣어 구현을 했지만, 추후 토큰 인증을 수행하고 만료 시간 확인 후 토큰 삭제 등의 인증 로직을 고려해봤을 때 아예 별도의 service로 분리하는 것은 어떨까요? 관심사 분리와 기능 확장에 더 유리할 것 같습니다 :)

2eve2 commented on src/main/java/com/flab/livecommerce/application/UserLoginProcessor.java:
```diff
@@ -0,0 +1,56 @@
+            throw new IllegalStateException();
+        }
+
+        //TODO 토큰 만료 LocalDateTime 을 어디에 추가할지, 로그인 요청이 계속 들어오면?
+        String token = tokenGenerator.generate();
+        tokenRepository.save(token, user);
+
+        return token;
```
현재는 login 서비스(프로세서) 로직 안에 토큰 생성과 저장 로직을 넣어 구현을 했지만, 추후 토큰 인증을 수행하고 만료 시간 확인 후 토큰 삭제 등의 인증 로직을 고려해봤을 때 아예 별도의 service로 분리하는 것은 어떨까요? 관심사 분리와 기능 확장에 더 좋을 것 같습니다 :)

dding94 replied:
추후 로직 고려하면 분리하는게 더 좋은 구조가 될 수 있겠네요!

dding94 replied:
다시 생각해봤는데 이게 login 이라는 프로세스에서 분리가 되는게 맞을까? 라는 생각이 드네요.
로그인이라는 한 과정안에서 작업이 이루어져야 할 것 같기도 합니다. 고민을 좀 해봐야 겠네요

2eve2 replied:
토큰 생성, 저장뿐만 아니라 추후 토큰을 삭제(만료)하거나 토큰저장소의 값을 비교하는 등의 확장성을 생각해보았을 때 이를 loginProcessor에서 처리를 해버리는 것보다 따로 분리해서 처리하는게 더 좋다고 판단했습니다. 저희가 레포지토리를 따로 사용하기도 하고요 :)
컨트롤러의 로그인 메서드 안에서는 'user 로그인 체크'와 '토큰생성 및 저장'이 명시되어 있어서 비즈니스 로직 처리 부분(서비스)만 나눠도 좋을 것 같은데, 좀 더 서로 얘기해봐야겠네요!

2eve2 committed:
refactor: 이메일 중복검사 API

GetMapping에서 PostMapping 방식으로 변경

2eve2 committed:
refactor: UserToken Service 분리

UserTokenManager로 비즈니스 로직 분리, 추후 인증 구현시 필요하다면 Processor 도입 고려

2eve2 committed:
fix: UserTest 수정

dding94 commented on src/main/java/com/flab/livecommerce/presentation/dto/InputEmail.java:
```diff
@@ -0,0 +1,19 @@
+package com.flab.livecommerce.presentation.dto;
```
dto패키지 따로 만드셨는데 request 패키지에 같이 있는게 좋지 않을까요?

2eve2 replied:
아 네 request에 합쳐도 좋을 것 같네요!

dding94 commented on src/main/java/com/flab/livecommerce/infrastructure/config/ProcessorConfig.java:
```diff
@@ -31,14 +32,19 @@ public UserCreateProcessor userCreateProcessor(
+        );
+    }
+
+    @Bean
+    public UserTokenManager userTokenManager(TokenRepository tokenRepository) {
+        return new UserTokenManager(
+            tokenRepository,
+            new NonInfoTokenGenerator()
```
UserTokenManager 를 bean 으로 등록한 이유가 있을까요??

2eve2 replied:
@Service로 등록하고 토큰 제너레이터만 빈으로 등록하도록 수정해야겠습니다

dding94 committed:
refactor: UserTokenManager 리팩토링

UserTokenManager 는 @Service 사용하고 Config에서는 TokenGenerator 를 빈으로 등록

dding94 committed:
refactor: email 중복체크 이름,패키지 변경
